### ISSUE TITLE
(exp): Compare CUDA versions

### ISSUE THREAD

**Issue by eric-czech** [kuleshov-group/dllm-dev#58]
I ran a quick experiment with the two Docker images mentioned in https://github.com/kuleshov-group/dllm-dev/pull/57.  One uses CUDA 12.4 and the other CUDA 12.8 (among other changes).  I wanted to see if any obvious problems pop up when making that upgrade.  I ran a script based on [bash_scripts/run_train_e2d2_wmt_lambda.sh](https://github.com/kuleshov-group/dllm-dev/blob/b569c703fc89f8fa7051d0cf3ec2bcbbfe76bf03/bash_scripts/run_train_e2d2_wmt_lambda.sh) to test them.

There aren't any big differences I saw in training throughput or loss between the two except for this odd throughput dip on 12.8 around 3k steps:

<img width="1024" alt="Image" src="https://github.com/user-attachments/assets/acce931c-2541-4a38-a804-f0f62ab4378a" />

I'm not sure what to make of that yet.  I'll look out for it as I keep running experiments, otherwise it doesn't seem like there are any clear reasons that upgrading Torch and CUDA won't work (yet).  @yair-schiff are you aware of any?

---

<details><summary>task.sky.yaml</summary>

```bash
workdir: .
run: |
  # Get the number of available GPUs
  NUM_VISIBLE_DEVICES=$(nvidia-smi -L | wc -l)
  
  # For Torch 2.7.0 and CUDA 12.8
  export TAG=e2d2_torch270
  export DOCKER_IMAGE=eczech/dllm-dev:py311-cuda128-torch270

  # For Torch 2.5.1 and CUDA 12.4
  # export TAG=e2d2_torch251
  # export DOCKER_IMAGE=eczech/dllm-dev:py312-cuda124-torch251

  sudo docker run \
    --env-file ~/.env \
    -e RUN_DIR=/workspace/outputs \
    -e DATA_DIR=/workspace/data \
    -e NUM_VISIBLE_DEVICES=$NUM_VISIBLE_DEVICES \
    -e HF_HOME=/workspace/.hf_cache \
    -e PYTHONPATH=/workspace:/workspace/.hf_cache/modules \
    -e TAG=$TAG \
    --gpus all --rm -v $(pwd):/workspace \
    $DOCKER_IMAGE \
    bash bash_scripts/run_train_e2d2_wmt_sky.sh
```

</details>

<details><summary>run_train_e2d2_wmt_sky.sh</summary>

```bash
#!/bin/bash

set -exo pipefail

# Fail fast on missing API keys
if [ -z "$WANDB_API_KEY" ]; then
    echo "Error: WANDB_API_KEY not set."
    exit 1
fi
if [ -z "$HUGGING_FACE_HUB_TOKEN" ]; then
    echo "Error: HUGGING_FACE_HUB_TOKEN not set."
    exit 1
fi

# Important variables (fix during hyperparam sweep)
BLOCK_SIZE=4
EVAL_BLOCK_SIZE=4
HIDDEN_SIZE=512 #512
INTERMEDIATE_SIZE=1536 #1536 #$(( 4 * HIDDEN_SIZE ))
N_ENCODER_LAYERS=28
ENCODER_TOP_LAYERS=false
N_DECODER_LAYERS=4
DECODER_TOP_LAYERS=false
REINIT_ENCODER=true
REINIT_DECODER=true
TIE_WEIGHTS=false
LOGIT_SHIFT=false
ENCODER_CAUSAL_MASK=false

# Hyperparameters
LR=3e-4
WARMUP_DURATION="1000ba"
BATCH_SIZE=128
MAX_DURATION="1000000ba"

PRETRAINED_MODEL_NAME_OR_PATH=Qwen/Qwen3-0.6B-Base

if [ "${ENCODER_TOP_LAYERS}" == "true" ]; then
  ENC_LAYERS="TOPenc${N_ENCODER_LAYERS}"
else
  ENC_LAYERS="enc${N_ENCODER_LAYERS}"
fi
if [ "${DECODER_TOP_LAYERS}" == "true" ]; then
  DEC_LAYERS="TOPdec${N_DECODER_LAYERS}"
else
  DEC_LAYERS="dec${N_DECODER_LAYERS}"
fi
RUN_NAME=wmt_block${BLOCK_SIZE}_lr${LR}_bsz${BATCH_SIZE}_warm${WARMUP_DURATION}_max-dur${MAX_DURATION}_${ENC_LAYERS}_${DEC_LAYERS}_hidden${HIDDEN_SIZE}_inter${INTERMEDIATE_SIZE}_${TAG}
if [ "${TIE_WEIGHTS}" == "true" ]; then
  RUN_NAME="${RUN_NAME}_tie-weights"
fi
if [ "${LOGIT_SHIFT}" == "true" ]; then
  RUN_NAME="${RUN_NAME}_logit-shift"
fi
if [ "${ENCODER_CAUSAL_MASK}" == "true" ]; then
  RUN_NAME="${RUN_NAME}_encoder-causal-mask"
fi
MICRO_BATCH_SIZE=16 #$(( BATCH_SIZE / NUM_VISIBLE_DEVICES ))
NUM_WORKERS=0

composer -n ${NUM_VISIBLE_DEVICES} scripts/composer_scripts/train_discrete_denoiser.py \
  run_name=${RUN_NAME} \
  pretrained_model_name_or_path=${PRETRAINED_MODEL_NAME_OR_PATH} \
  dataset@train_dataset=wmt_train \
  dataset@eval_dataset=wmt_eval \
  composer.optimizer.lr=${LR} \
  composer.trainer.save_num_checkpoints_to_keep=1 \
  composer.trainer.eval_interval="5000ba" \
  composer.trainer.max_duration=${MAX_DURATION} \
  composer.loggers.name=${RUN_NAME} \
  composer/lr_scheduler=constant_with_warmup \
  composer.lr_scheduler.t_warmup=${WARMUP_DURATION} \
  model=e2d2 \
  model.config.attn_backend="sdpa" \
  training.compile_backbone=true \
  model.config.length=256 \
  model.config.shift_logits=${LOGIT_SHIFT} \
  model/backbone@model.config.backbone_config=llm_as_encoder_decoder \
  model.config.backbone_config.use_encoder_causal_mask=${ENCODER_CAUSAL_MASK} \
  model.config.backbone_config.num_encoder_layers=${N_ENCODER_LAYERS} \
  model.config.backbone_config.num_decoder_layers=${N_DECODER_LAYERS} \
  model.config.backbone_config.tie_encoder_decoder_weights=${TIE_WEIGHTS} \
  model.config.backbone_config.reinit_decoder=${REINIT_DECODER} \
  model.config.backbone_config.reinit_encoder=${REINIT_ENCODER} \
  model.config.backbone_config.keep_top_decoder_layers=${DECODER_TOP_LAYERS} \
  model.config.backbone_config.keep_top_encoder_layers=${ENCODER_TOP_LAYERS} \
  +model.config.backbone_config.hidden_size=${HIDDEN_SIZE} \
  +model.config.backbone_config.intermediate_size=${INTERMEDIATE_SIZE} \
  training.global_batch_size=${BATCH_SIZE} \
  training.grad_accum=$(( BATCH_SIZE / NUM_VISIBLE_DEVICES / MICRO_BATCH_SIZE )) \
  ~composer.trainer.compile_config \
  ~composer.trainer.parallelism_config \
  block_size=${BLOCK_SIZE} \
  eval_block_size=${EVAL_BLOCK_SIZE} \
  training.antithetic_sampling=false \
  hydra.run.dir=${RUN_DIR}/${RUN_NAME} \
  composer.trainer.save_interval="1000ba" \
  train_dataloader.num_workers=${NUM_WORKERS} \
  composer.callbacks.hf_compatible_checkpointing.disable_hf=true \
  composer.callbacks.speed_monitor.window_size=10

```

</details>

---

W&B runs:

- [CUDA 12.4 run](https://wandb.ai/eric-czech/discrete-diffusion/runs/wmt_block4_lr3e-4_bsz128_warm1000ba_max-dur1000000ba_enc28_dec4_hidden512_inter1536_e2d2_torch251)
- [CUDA 12.8 run](https://wandb.ai/eric-czech/discrete-diffusion/runs/wmt_block4_lr3e-4_bsz128_warm1000ba_max-dur1000000ba_enc28_dec4_hidden512_inter1536_e2d2_torch270)
